woeip/apps/air_quality/views.py

```diff
@@ -0,0 +1,66 @@
+import logging
+
+from django.contrib import messages
+from django.contrib.auth.decorators import login_required
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import Error
+from django.shortcuts import redirect, render
+from django.utils.encoding import force_text
+
+from woeip.apps.air_quality import dustrak, forms, models
+
+logger = logging.getLogger(__name__)
+
+
+def index(request):
+    return render(request, 'index.html')
+
+
+@login_required
+def upload(request):
+    """Upload data for a session collected using the Dustrak air quality device and a separate GPS
+    log file. Creates a new Session instance, two SessionData instances, and Data instances for
+    each sample.
+    """
+
+    if request.method == 'POST':
+        form = forms.DustrakSessionForm(request.POST, request.FILES)
+        if form.is_valid():
+            form.save()
+            try:
+                air_sensor = models.Sensor.objects.get(name='Dustrak')
+                gps_sensor = models.Sensor.objects.get(name='GPS')
+
+                air_quality = models.SessionData(upload=request.FILES['air_quality'],
+                                                 sensor=air_sensor,
+                                                 session=form.instance,
+                                                 uploaded_by=request.user)
+                gps = models.SessionData(upload=request.FILES['gps'],
+                                         sensor=gps_sensor,
+                                         session=form.instance,
+                                         uploaded_by=request.user)
+
+                air_quality_contents = force_text(request.FILES['air_quality'].read())
+                _, air_quality_data = dustrak.load_dustrak(air_quality_contents, 'America/Los_Angeles')
+
+                gps_contents = force_text(request.FILES['gps'].read())
+                gps_data = dustrak.load_gps(gps_contents)
+                joined_data = dustrak.join(air_quality_data, gps_data)
+
+            except (UnboundLocalError, ObjectDoesNotExist, Error) as e:
```

Comment:
What causes these errors? I have never seen `UnboundLocalError`. Do some validation beforehand to catch `ObjectDoesNotExist` and return an error sooner rather than later. `Error` is way too broad. Rather than catch it, let it blow up.

The higher level question I ask myself is, "what is the minimum amount of code that needs to be in this try/except block"?

Reply:
The errors are actually related. `UnboundLocalError` is caused when the `air_sensor` and `gps` variables are called without being assigned a value. They fail to be assigned when there is an `ObjectDoesNotExist`. I've struggled to find guidance on good 'try/except' methodology. This comment helps a lot.

Reply:
After thinking it over, my understanding for exception handling is:
1) Understand what can go wrong
2) Code defensively, to keep it from going wrong
3) Add exception handling to catch known errors that are difficult to prevent/out of the developers control

As it applies to this case, the only known error is related to loading the dustrak and gps sensor information. Because the process is not yet understood, it's impractacle to code defensively/add validation. Consequently, I modified the code to catch the known error and let the rest of the view 'blow up'. Validation will be added in a later PR, once the dustrak/gps upload process is established.